staging/src/k8s.io/api/core/v1/types.go

```diff
@@ -624,6 +624,29 @@ type GlusterfsVolumeSource struct {
 	ReadOnly bool `json:"readOnly,omitempty" protobuf:"varint,3,opt,name=readOnly"`
 }
 
+// Represents a Glusterfs mount that lasts the lifetime of a pod.
+// Glusterfs volumes do not support ownership management or SELinux relabeling.
+type GlusterfsPersistentVolumeSource struct {
+	// EndpointsName is the endpoint name that details Glusterfs topology.
+	// More info: https://releases.k8s.io/HEAD/examples/volumes/glusterfs/README.md#create-a-pod
+	EndpointsName string `json:"endpoints" protobuf:"bytes,1,opt,name=endpoints"`
```

Comment:
IIUC the fields name are kept the same as `GlusterfsVolumeSource` for some sort of backwards compatibility and symmetry. But it would be backwards incompatible if someone expects `PersistentVolumeSource.Glusterfs` to be `GlusterfsVolumeSource` (which turns out to be `GlusterfsPersistentVolumeSource` now). 

In that case should we fix the go name (s/EndpointsName/Endpoints) now to follow the api rule https://github.com/kubernetes/kubernetes/blob/master/vendor/k8s.io/kube-openapi/pkg/generators/rules/names_match.go#L53-L78? 

Also FYI I'm planning on fixing the existing go name violations (e.g. https://github.com/kubernetes/kubernetes/pull/66097). The fix will be breaking change, but it's mechanical change and will be caught by compiler.

Reply:
> But it would be backwards incompatible if someone expects PersistentVolumeSource.Glusterfs to be GlusterfsVolumeSource (which turns out to be GlusterfsPersistentVolumeSource now).

Not via the API. The serialized JSON and protobuf output is identical, no matter what struct it comes from.

> In that case should we fix the go name (s/EndpointsName/Endpoints) now to follow the api rule https://github.com/kubernetes/kubernetes/blob/master/vendor/k8s.io/kube-openapi/pkg/generators/rules/names_match.go#L53-L78?

No, we must maintain wire compatibility with all the fields of the old struct.

> Also FYI I'm planning on fixing the existing go name violations (e.g. #66097)

That is ok as long as the serialized wire formats do not change in any way.

Reply:
Yes the change is backwards compatible via the API. I'm thinking if someone builds a go client that manipulates the field `Glusterfs` as a `GlusterfsVolumeSource` type, e.g.
```
    pvs := PersistentVolumeSource{
        Glusterfs: &GlusterfsVolumeSource{}
    }
```
then it will be a breaking change when the user updates `k8s.io/api/core/v1` package, because `Glusterfs` should be of type `GlusterfsPersistentVolumeSource` now. Thus I'm thinking we could also fix the go field name `EndpointsName` to be `Endpoints` as part of the breaking change. Sorry I guess my previous comment was misleading, I agree we shouldn't change the json field name (the serialized wire formats), but we could change the go field name.

Reply:
> Thus I'm thinking we could also fix the go field name EndpointsName to be Endpoints as part of the breaking change. 

leaving the go field names the same means a client that obtains a PV object from client.Core().PersistentVolumes().Get("...") would not need to make changes